File: Algorithms/src/712. Minimum ASCII Delete Sum for Two Strings/Code/solution.py
class Solution:
    def minimumDeleteSum(self, s1: str, s2: str) -> int:
        n, m = len(s1), len(s2)
        dp = [0] * (m + 1)

        for j in range(m - 1, -1, -1):
            dp[j] = dp[j + 1] + ord(s2[j])

        for i in range(n - 1, -1, -1):
            prev = dp[m]
            dp[m] += ord(s1[i])

            for j in range(m - 1, -1, -1):
                temp = dp[j]
                if s1[i] == s2[j]:
                    dp[j] = prev
                else:
                    dp[j] = min(
                        ord(s1[i]) + dp[j],
                        ord(s2[j]) + dp[j + 1]
                    )
                prev = temp

        return dp[0]

File: Algorithms/src/712. Minimum ASCII Delete Sum for Two Strings/Code/solution.cpp
class Solution {
public:
    int minimumDeleteSum(string s1, string s2) {
        int n = s1.size(), m = s2.size();
        vector<int> dp(m + 1, 0);

        // Base case: when s1 is empty
        for (int j = m - 1; j >= 0; j--) {
            dp[j] = dp[j + 1] + s2[j];
        }

        for (int i = n - 1; i >= 0; i--) {
            int prev = dp[m];          // dp[i+1][m]
            dp[m] += s1[i];            // delete remaining chars of s1

            for (int j = m - 1; j >= 0; j--) {
                int temp = dp[j];
                if (s1[i] == s2[j]) {
                    dp[j] = prev;
                } else {
                    dp[j] = min(
                        s1[i] + dp[j],
                        s2[j] + dp[j + 1]
                    );
                }
                prev = temp;
            }
        }
        return dp[0];
    }
};
